File: tools/tex-compile-vps/app/errori_tex.py
"""Gli errori di pdflatex, letti dal suo log (24/9/2026).

pdflatex in `nonstopmode` non si ferma al primo errore: lo scrive nel log,
prova a rimediare e spesso produce lo stesso un PDF, con la figura a metà.
Fino al 24/9/2026 `render_tikz` giudicava solo dall'esistenza del PDF: una
chiave TikZ sbagliata o un `\\foreach` rotto davano `ok=True` e un SVG
parziale, e al docente non arrivava nessun messaggio. Quando invece il PDF
non usciva, il log restituito erano i primi e gli ultimi 2000 caratteri: il
caricamento dei pacchetti, e l'errore solo se stava in fondo.

Qui: quali errori ci sono (ogni riga che comincia con `! `, come li scrive
TeX) e un estratto leggibile che comincia da loro. Solo la libreria standard,
per poterlo provare senza installare il servizio (`tests/tex/`).
"""
from __future__ import annotations

import re

# `l.136 ...testo della riga` — la riga del sorgente a cui TeX si è fermato.
_RIGA = re.compile(r"^l\.(\d+)")

# Le righe di servizio fra il messaggio e `l.NNN`: vuote, i puntini con cui
# TeX abbrevia, l'invito a premere H (che in nonstopmode non si può).
_RUMORE = re.compile(r"^\s*(\.\.\.)?\s*$|^Type\s+H <return>|^See the \S+ package documentation")

# La riga con cui pdflatex chiude dopo un errore fatale: non è un errore in
# più, è la conseguenza del primo.
_CHIUSURE = ("==> Fatal error occurred", "Emergency stop")


def trova_errori(log: str, limite: int = 20) -> list[dict]:
    """Gli errori del log, nell'ordine in cui compaiono.

    Ognuno: `message` (la riga dopo `! `), `line` (il numero dopo `l.`, se
    TeX l'ha scritto entro poche righe; è la riga del documento compilato,
    che può avere qualche riga in più del sorgente del docente) e `context`
    (le righe fra il messaggio e `l.NNN`, più quella dopo: TeX spezza la
    riga del sorgente nel punto esatto in cui si è fermato).
    """
    errori: list[dict] = []
    if not log:
        return errori
    righe = log.splitlines()
    for i, riga in enumerate(righe):
        if not riga.startswith("! "):
            continue
        messaggio = riga[2:].strip()
        if messaggio.startswith(_CHIUSURE) and errori:
            continue
        numero = None
        contesto: list[str] = []
        for j in range(i + 1, min(i + 14, len(righe))):
            seguente = righe[j]
            if seguente.startswith("! "):
                break
            trovata = _RIGA.match(seguente)
            if not trovata and _RUMORE.match(seguente):
                continue
            contesto.append(seguente)
            if trovata:
                numero = int(trovata.group(1))
                # TeX spezza la riga nel punto in cui si è fermato: la
                # seconda metà sta nella riga dopo, rientrata.
                if j + 1 < len(righe) and righe[j + 1].strip():
                    contesto.append(righe[j + 1])
                break
        errori.append({
            "line": numero,
            "message": messaggio[:300],
            "context": "\n".join(contesto).strip("\n")[:800],
        })
        if len(errori) >= limite:
            break
    return errori


def estratto(log: str, errori: list[dict], max_chars: int = 4000) -> str:
    """Il testo da mostrare al docente: prima gli errori con il loro
    contesto, poi la coda del log (dove pdflatex dice come è finita)."""
    if not errori:
        return log if len(log) <= max_chars else "[...]\n" + log[-max_chars:]
    n = len(errori)
    parti = [f"pdflatex ha trovato {n} {'errore' if n == 1 else 'errori'}:"]
    for e in errori:
        blocco = "! " + e["message"]
        if e.get("context"):
            blocco += "\n" + e["context"]
        parti.append(blocco)
    testa = "\n\n".join(parti)
    resto = max_chars - len(testa) - 40
    if resto > 200:
        coda = log[-resto:] if len(log) > resto else log
        testa += "\n\n--- fine del log di pdflatex ---\n" + coda
    return testa[: max_chars + 200]

File: app/Services/TexCompile/ErroriTex.php
<?php

declare(strict_types=1);

namespace App\Services\TexCompile;

/**
 * Gli errori di pdflatex letti dal suo log, e l'estratto da mostrare al
 * docente (24/9/2026).
 *
 * pdflatex in nonstopmode rimedia agli errori e produce quasi sempre un PDF:
 * il servizio TeX lo dava per riuscito e la figura arrivava a metà, senza
 * messaggio. Gemello di `tools/tex-compile-vps/app/errori_tex.py`, che fa lo
 * stesso lato servizio per l'anteprima SVG; qui serve al modal, a cui il
 * servizio manda il log intero (`/compile?with_artifacts=1`). Le due copie
 * devono dire la stessa cosa: le provano `tests/tex/test_errori_tex.py` e
 * `tests/Unit/Services/TexCompile/ErroriTexTest.php` sugli stessi log.
 */
final class ErroriTex
{
    /**
     * Ogni riga che comincia con `! ` è un errore di TeX. `line` è il numero
     * dopo `l.` (riga del documento compilato), `context` le righe fino a
     * `l.NNN` più quella dopo, dove TeX mostra il resto della riga.
     *
     * @return list<array{line:?int, message:string, context:string}>
     */
    public static function daLog(string $log, int $limite = 20): array
    {
        $errori = [];
        if ($log === '') {
            return $errori;
        }
        $righe = preg_split('/\r\n|\n|\r/', $log) ?: [];
        $n = \count($righe);
        foreach ($righe as $i => $riga) {
            if (!str_starts_with($riga, '! ')) {
                continue;
            }
            $messaggio = trim(substr($riga, 2));
            if ($errori !== [] && (str_starts_with($messaggio, '==> Fatal error occurred') || str_starts_with($messaggio, 'Emergency stop'))) {
                continue;
            }
            $numero = null;
            $contesto = [];
            for ($j = $i + 1; $j < min($i + 14, $n); $j++) {
                $seguente = $righe[$j];
                if (str_starts_with($seguente, '! ')) {
                    break;
                }
                $trovata = preg_match('/^l\.(\d+)/', $seguente, $m) === 1;
                if (!$trovata && preg_match('/^\s*(\.\.\.)?\s*$|^Type\s+H <return>|^See the \S+ package documentation/', $seguente) === 1) {
                    continue;
                }
                $contesto[] = $seguente;
                if ($trovata) {
                    $numero = (int) $m[1];
                    if ($j + 1 < $n && trim($righe[$j + 1]) !== '') {
                        $contesto[] = $righe[$j + 1];
                    }
                    break;
                }
            }
            $errori[] = [
                'line'    => $numero,
                'message' => mb_substr($messaggio, 0, 300),
                'context' => mb_substr(trim(implode("\n", $contesto), "\n"), 0, 800),
            ];
            if (\count($errori) >= $limite) {
                break;
            }
        }
        return $errori;
    }

    /**
     * Gli errori come li manda il servizio TeX (`errori_tex.trova_errori`),
     * ridotti alla forma di `daLog`: si tiene solo quel che ha la forma
     * giusta, e i testi con i loro tetti.
     *
     * @param array<mixed> $dalServizio
     * @return list<array{line:?int, message:string, context:string}>
     */
    public static function normalizza(array $dalServizio): array
    {
        $errori = [];
        foreach ($dalServizio as $e) {
            if (!\is_array($e) || !\is_string($e['message'] ?? null) || $e['message'] === '') {
                continue;
            }
            $errori[] = [
                'line'    => \is_int($e['line'] ?? null) ? $e['line'] : null,
                'message' => mb_substr($e['message'], 0, 300),
                'context' => mb_substr(\is_string($e['context'] ?? null) ? $e['context'] : '', 0, 800),
            ];
            if (\count($errori) >= 20) {
                break;
            }
        }
        return $errori;
    }

    /**
     * Prima gli errori con il loro contesto, poi la coda del log (dove
     * pdflatex dice com'è finita). Senza errori: la coda del log.
     *
     * @param list<array{line:?int, message:string, context:string}> $errori
     */
    public static function estratto(string $log, array $errori, int $max = 8000): string
    {
        if ($errori === []) {
            return \strlen($log) <= $max ? $log : "[...]\n" . substr($log, -$max);
        }
        $n = \count($errori);
        $parti = ['pdflatex ha trovato ' . $n . ($n === 1 ? ' errore:' : ' errori:')];
        foreach ($errori as $e) {
            $parti[] = '! ' . $e['message'] . ($e['context'] !== '' ? "\n" . $e['context'] : '');
        }
        $testa = implode("\n\n", $parti);
        $resto = $max - \strlen($testa) - 40;
        if ($resto > 200) {
            $testa .= "\n\n--- fine del log di pdflatex ---\n" . (\strlen($log) > $resto ? substr($log, -$resto) : $log);
        }
        return substr($testa, 0, $max + 200);
    }
}
